frontend/catalyst/cuda/primitives/__init__.py

```diff
@@ -0,0 +1,578 @@
+        We can just say that multiple_results = True and return an empty tuple.
+        """
+        kernel_gate_p.bind(kernel, *qubits_or_params, inst=inst, qubits_len=qubits_len)
+        return tuple()
+
+    @kernel_gate_p.def_impl
+    def gate_impl(kernel, *qubits_or_params, inst=None, qubits_len=-1):
+        """Concrete implementation."""
+        assert inst and qubits_len > 0
+        method = getattr(cudaq.Kernel, inst)
+        targets = qubits_or_params[:qubits_len]
+        params = qubits_or_params[qubits_len:]
+        method(kernel, *params, *targets)
+        return tuple()
+
+    @kernel_gate_p.def_abstract_eval
+    def gate_abs(_kernel, *_qubits_or_params, inst=None, qubits_len=-1):
+        """Abstract evaluation."""
+        return tuple()
+
+    return gate_func, kernel_gate_p
+
+
+cuda_inst, cuda_inst_p = make_primitive_for_gate()
+
+
+def make_primitive_for_m(gate: str):
+    """A single function to make primitives for all measurement basis.
+
+    Args:
+      * gate (str): Either "x", "y", or "z" are valid values.
+    Returns:
+      * A function that binds a primitive.
+      * the primitive itself.
+    """
+
+    assert gate in {"x", "y", "z"}
+    gate = f"m{gate}"
+    kernel_gate_p = jax.core.Primitive(f"kernel_{gate}")
+    method = getattr(cudaq.Kernel, gate)
+
+    def gate_func(kernel, target):
+        """Convenience."""
+        return kernel_gate_p.bind(kernel, target)
+
+    @kernel_gate_p.def_impl
+    def gate_impl(kernel, target):
+        """Concrete implementation."""
+        return method(kernel, target)
+
+    @kernel_gate_p.def_abstract_eval
+    def gate_abs(_kernel, _target):
+        """Abstract evaluation."""
+        return AbsCudaValue()
+
+    return gate_func, kernel_gate_p
+
+
+mx_call, mx_p = make_primitive_for_m("x")
+my_call, my_p = make_primitive_for_m("y")
+mz_call, mz_p = make_primitive_for_m("z")
+
+
+cudaq_sample_p = jax.core.Primitive("cudaq_sample")
+cudaq_counts_p = jax.core.Primitive("cudaq_counts")
+cudaq_counts_p.multiple_results = True
+
+
+def cudaq_sample(kernel, *args, shots_count=1000):
+    """Convenience function for binding."""
+    return cudaq_sample_p.bind(kernel, *args, shots_count=shots_count)
+
+
+@cudaq_sample_p.def_impl
+def cudaq_sample_impl(kernel, *args, shots_count=1000):
+    """Concrete implementation of cudaq.sample.
+
+    `cudaq.sample` returns an object which is a compressed version of what
+    `qml.sample` returns as samples. Instead of returning an array with the observed
+    population, `cudaq.sample` returns a dictionary where the keys are bitstrings and
+    values are the frequency those bitstrings were observed.
+
+    In a way `qml.count` is more similar to `cudaq.sample` than `qml.sample`.
+    So, let's perform a little conversion here.
+    """
+    a_dict = cudaq.sample(kernel, *args, shots_count=shots_count)
+    lls = [[k] * v for k, v in a_dict.items()]
+    return [l for ls in lls for l in ls]
```

Comment:
should we update the return value to jnp.array? :thinking:

Reply:
https://github.com/PennyLaneAI/catalyst/pull/477/commits/7a8da712deaceca34c13f84efb3a8af8f6762c67